### PR TITLE
Pd mix substeps

### LINKED ISSUE #725: PD Basic Pipetting Steps: Mix (Sidebar)
As a protocol designer, I'd like to be able to view created mix steps in the steps sidebar.

## Acceptance Criteria
- Visual design updated to reflect mockup (linked below)
- User can view selected ingredient
- User can view icon representing step
- Clicking on step or sub-step in sidebar opens mix editor

image

- Show labware name, volume (eg `10uL`), and repetitions (eg `8x`)
- 1 row per well

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

Closes #725 and fixes a bug where updated liquid state wasn't being shown in the substeps (eg if you transferred liquid to an empty well in Step 1, then dispensed from that well, the dispense subcommands would look like you were aspirating from an empty well. Fixed in this PR.)

## changelog

- Complib: minor type definition changes to `TitledList`
- Mix substep header: `protocol-designer/src/components/steplist/MixHeader.js`
- Mix substep data generated:  `protocol-designer/src/steplist/generateSubsteps.js`
- Steps have numbered names eg `Transfer 1`

The majority of the diff is refactoring `StepList` & `StepItems` to try and clean it up:
* Container for `StepItem` components, before StepList was responsible for passing down all props
* `ConnectedStepList` is now a very minimal container since `ConnectedStepItem` gets the props it needs by itself
* Break out `protocol-designer/src/components/StepList/index.js` into several small components
* Selectors: instead of mixing in `substeps` data in the `allStepsWithSubsteps` selector, just select `allSteps` and `allSubsteps` separately. Merging them just made things more complicated.
* `SourceDestSubstep` file had several components; broken up into their own files

## review requests

- Make sure Mix subcommands work fully as expected, including hover-to-highlight-well behavior, etc
- Confirm the subcommands liquid state bug explained in "overview" is fixed
- The code changes are largely just breaking things apart into their own files, and connecting `StepItem`s via their container.

### CODE REVIEW

File: components/src/lists/TitledList.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ type ListProps = {
   /** text of title */
   title: string,
   /** optional icon left of the title */
-  iconName?: IconName,
+  iconName?: ?IconName,
```
Comment (IanLondon) [inline components/src/lists/TitledList.js]: NOTE: minor type def changes to TitledList

File: protocol-designer/src/containers/ConnectedStepItem.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+// @flow
```
Comment (IanLondon) [inline protocol-designer/src/containers/ConnectedStepItem.js]: This container is new

File: protocol-designer/src/components/steplist/MixHeader.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// @flow
```
Comment (IanLondon) [inline protocol-designer/src/components/steplist/MixHeader.js]: new "substep header" for Mix

File: protocol-designer/src/steplist/generateSubsteps.js [hunk 4]
```diff
@@ -194,11 +195,62 @@ function _consolidateSubsteps (
 function _distributeSubsteps (
   form: DistributeFormData,
   transferLikeFields: *
-): ?TransferLikeSubstepItem { // <-- TODO remove '?' type
+): ?SourceDestSubstepItem { // <-- TODO remove '?' type
   console.log('Distribute substeps not yet implemented') // TODO Ian 2018-05-04
   return null
 }
```
Comment (IanLondon) [inline protocol-designer/src/steplist/generateSubsteps.js]: mix substeps. Going to try and merge the parts of substep generation that overlap btw transferlike step types, and mix, in the Distribute Substeps PR

File: protocol-designer/src/containers/ConnectedStepList.js [hunk 5]
```diff
@@ -1,46 +1,17 @@
 // @flow
```
Comment (IanLondon) [inline protocol-designer/src/containers/ConnectedStepList.js]: This container pared down greatly, most of the functionality moved to ConnectedStepItem

File: protocol-designer/src/containers/ConnectedStepItem.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import {connect} from 'react-redux'
+import type {BaseState, ThunkDispatch} from '../types'
+
+import type {SubstepIdentifier} from '../steplist/types'
+import {hoverOnSubstep, selectStep, hoverOnStep, toggleStepCollapsed} from '../steplist/actions'
+import * as substepSelectors from '../top-selectors/substeps'
+import {selectors as steplistSelectors} from '../steplist/reducers'
+import {selectors as fileDataSelectors} from '../file-data'
+import {selectors as labwareIngredSelectors} from '../labware-ingred/reducers'
+import StepItem from '../components/steplist/StepItem' // TODO Ian 2018-05-10 why is importing StepItem from index.js not working?
```
Comment (IanLondon) [inline protocol-designer/src/containers/ConnectedStepItem.js]: It confuses me why `import {StepItem} from '../components/steplist'` isn't working, but `import {StepList} from '../components/steplist'` works fine in `ConnectedStepList`

**Comment by Kadee80** [comment]
Did an initial pass, not sure whats in scope.

- Went to make a mix step, decided to delete it (in more options). I remembered the warning about not having a tip if you try and mix first. Got a white screen after clicking delete.

- Did a distribute from one tube in a tube rack to 8 wells in a plate. (with p10 single) when I hover over that step, whole plate, not individual wells are highlighted.

- Mixed the first column of the 96 well plate with all the bells and whistles (with p10 multi) and success! wells highlighted etc. 
Question: does the multi (which I had not used yet in the protocol) know to pick up tips if it does not have any? I remember getting warned that I had no tip when trying to do a mix command as the first step earlier in the week. This was not the first step, but it was the first step with the multi channel.

**Pull request merged**